File: cpp/problems/leetcode/0647_palindromic_substrings/solution.cpp
// LeetCode 647: Palindromic Substrings
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int countSubstrings(string s) {
        int n = (int)s.size(), ans = 0;
        auto expand = [&](int l, int r) {
            int cnt = 0;
            while (l >= 0 && r < n && s[l] == s[r]) {
                ++cnt;
                --l;
                ++r;
            }
            return cnt;
        };
        for (int i = 0; i < n; ++i) {
            ans += expand(i, i);
            ans += expand(i, i + 1);
        }
        return ans;
    }
};

int main() {
    assert(Solution().countSubstrings("abc") == 3);
    assert(Solution().countSubstrings("aaa") == 6);
    assert(Solution().countSubstrings("") == 0);
    cout << "0647 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0647_palindromic_substrings/solution.py
"""LeetCode 647: Palindromic Substrings."""

from __future__ import annotations


class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        ans = 0

        def expand(l: int, r: int) -> int:
            cnt = 0
            while l >= 0 and r < n and s[l] == s[r]:
                cnt += 1
                l -= 1
                r += 1
            return cnt

        for i in range(n):
            ans += expand(i, i)
            ans += expand(i, i + 1)
        return ans


if __name__ == "__main__":
    assert Solution().countSubstrings("abc") == 3
    assert Solution().countSubstrings("aaa") == 6
    assert Solution().countSubstrings("") == 0
    print("0647 OK")
